Implement RFC 1861: Extern types

plietar opened the pull request:
A few notes :

- Type parameters are not supported. This was an unresolved question from the RFC. It is not clear how useful this feature is, and how variance should be treated. This can be added in a future PR.

- `size_of_val` / `align_of_val` can be called with extern types, and respectively return 0 and 1. This differs from the RFC, which specified that they should panic, but after discussion with @eddyb on IRC this seems like a better solution.
If/when a `DynSized` trait is added, this will be disallowed statically.

- Auto traits are not implemented by default, since the contents of extern types is unknown. This means extern types are `!Sync`, `!Send` and `!Freeze`. This seems like the correct behaviour to me.
Manual `unsafe impl Sync for Foo` is still possible.

- This PR allows extern type to be used as the tail of a struct, as described by the RFC :
```rust
extern {
    type OpaqueTail;
}

#[repr(C)]
struct FfiStruct {
    data: u8,
    more_data: u32,
    tail: OpaqueTail,
}
```

However this is undesirable, as the alignment of `tail` is unknown (the current PR assumes an alignment of 1). Unfortunately we can't prevent it in the general case as the tail could be a type parameter : 
```rust
#[repr(C)]
struct FfiStruct<T: ?Sized> {
    data: u8,
    more_data: u32,
    tail: T,
}
```

Adding a `DynSized` trait would solve this as well, by requiring tail fields to be bound by it.

- Despite being unsized, pointers to extern types are thin and can be casted from/to integers. However it is not possible to write a `null<T>() -> *const T` function which works with extern types, as I've explained here : https://github.com/rust-lang/rust/issues/43467#issuecomment-321678621

- Trait objects cannot be built from extern types. I intend to support it eventually, although how this interacts with `DynSized`/`size_of_val` is still unclear.

- The definition of `c_void` is unmodified

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @arielb1 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

plietar commented:
I also have an implementation of DynSized which goes with this at https://github.com/plietar/rust/tree/dynsized

arielb1 commented on src/librustc_trans/intrinsic.rs:
```diff
@@ -139,13 +139,15 @@ pub fn trans_intrinsic_call<'a, 'tcx>(bcx: &Builder<'a, 'tcx>,
             } else {
-                let lltp_ty = type_of::type_of(ccx, tp_ty);
-                C_uint(ccx, machine::llsize_of_alloc(ccx, lltp_ty))
+                C_null(llret_ty)
```
What is that supposed to be? Did you mean `C_uint(ccx, 0)`?

plietar replied:
Yes, although I'm not really familiar with LLVM IR and I expected the two to be equivalent. I originally took this from the `discriminant_value` intrinsic.

arielb1 commented on src/librustc_trans/mir/lvalue.rs:
```diff
@@ -145,10 +145,11 @@ impl<'a, 'tcx> LvalueRef<'tcx> {
-                return (bcx.struct_gep(
-                        ptr_val, adt::struct_llfields_index(st, ix)), alignment);
-            }
+            bcx.ccx.shared().type_is_sized(fty)
```
You don't need to do an alignment fixup if you have a foreign vtable. Actually, you only need to do an alignment fixup if the vtable type is `Trait`.

plietar replied:
Foreign tails were handled by the `!self.has_extra()` a few lines down. I've added a case to the match just beneath, especially since I'm pretty sure the `!self.has_extra()` branch should never be taken.

arielb1 replied:
I think it's detritus from old trans, but with your patch it's taken by `Foo<Foo<Opaque>>`. I'm fine with it because you're trying to forbid that.

arielb1 commented on src/test/compile-fail/extern-types-not-sync-send.rs:
```diff
@@ -0,0 +1,28 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
```
Also add a test that they are `!Sized`?

plietar replied:
Done

arielb1 commented on src/test/run-pass/extern-types-pointer-cast.rs:
```diff
@@ -0,0 +1,30 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
```
It would be better to use a pointer with an "interesting" bit pattern rather than `0`, to make sure we are not truncating anything.

plietar replied:
Done. I've had to use `cfg(target_pointer_width)` to make sure all bits have something interesting.

arielb1 commented on src/test/run-pass/extern-types-size_of_val.rs:
```diff
@@ -0,0 +1,26 @@
+
+    assert_eq!(size_of_val(x), 0);
+    assert_eq!(align_of_val(x), 1);
+}
```
More tests I think we need:

1) test that pointers to extern types have the same *size* and *align* as a thin pointer.
2) check that pointers to extern types have the right ABI - maybe add a test to https://github.com/rust-lang/rust/tree/e22a3cf7e12b622b143694942504c66a544dc4c5/src/test/run-make/extern-fn-struct-passing-abi
3) check that structs with `extern` tails work - e.g. `struct Foo<T: ?Sized>(u8, T);` where `T = <foreign type>/Foo<<foreign type>>`

arielb1 commented:
Basically LGTM modulo the 2 nits and the tests.

plietar commented:
@arielb1 I've addressed your comments.

> check that structs with extern tails work - e.g. struct Foo<T: ?Sized>(u8, T); where T = <foreign type>/Foo<<foreign type>>

I've changed most of the tests to try both extern types and struct with foreign tails.

I've however not put any test which tries to get the address of the foreign tail, since the alignment of the field is unknown. While it is currently allowed (with an assumed alignement of 1), it shouldn't be once DynSized is implemented.

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit 5814c88 has been approved by `arielb1`

<!-- @bors r=arielb1 5814c880fcf31abd4895295d14f6d994b2badf34 -->

bors commented:
:hourglass: Testing commit 5814c880fcf31abd4895295d14f6d994b2badf34 with merge 9dbee87171690a1188497154b292f6403992db03...

kennytm commented on src/libsyntax/feature_gate.rs:
```diff
@@ -382,6 +382,9 @@ declare_features! (
     (active, match_beginning_vert, "1.21.0", Some(44101)),
+
+    // extern types
+    (active, extern_types, "1.21.0", Some(43467)),
```
Shouldn't this be `1.22.0`?

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.4545)

kennytm commented:
`dist x86_64-pc-windows-msvc` failed, needs to update `rustfmt`. Legit.

```
[01:03:51] error[E0004]: non-exhaustive patterns: `Ty` not covered
[01:03:51]     --> C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\rustfmt-nightly-0.2.5\src\items.rs:2808:39
[01:03:51]      |
[01:03:51] 2808 |         let item_str = try_opt!(match self.node {
[01:03:51]      |                                       ^^^^^^^^^ pattern `Ty` not covered
[01:03:51] 
[01:03:51] error: aborting due to previous error
[01:03:51] 
[01:03:51] error: Could not compile `rustfmt-nightly`.
```

plietar commented:
Opened PR in rustfmt : https://github.com/rust-lang-nursery/rustfmt/pull/1946

plietar commented:
ping @arielb1, I've update rustfmt

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit d60d340 has been approved by `arielb1`

<!-- @bors r=arielb1 d60d340edf68e965d86b5262070b2d4624f8ffea -->

bors commented:
:hourglass: Testing commit d60d340edf68e965d86b5262070b2d4624f8ffea with merge abf1b10f2b7028b9c169abdae435f76e24397d2d...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/273759124?utm_source=github_status&utm_medium=notification)

kennytm commented:
Could not build rustdoc. Legit, but it seems unrelated to this PR.

```rust
[00:32:38] error[E0599]: no method named `as_str` found for type `syntax_pos::symbol::InternedString` in the current scope
[00:32:38]     --> /checkout/src/librustdoc/clean/mod.rs:1942:69
[00:32:38]      |
[00:32:38] 1942 |                 let path = external_path(cx, &cx.tcx.item_name(did).as_str(),
[00:32:38]      |                                                                     ^^^^^^
[00:32:38] 
[00:32:40] error: aborting due to previous error
```

bors commented:
:umbrella: The latest upstream changes (presumably #44316) made this pull request unmergeable. Please resolve the merge conflicts.

plietar commented:
Rebased

arielb1 commented:
Looking at the code again, I see that there are a few more cases where `TyForeign` should be treated like e.g. `TyAdt`. Could you try to make it a kind of adt?

plietar commented:
@arielb1 I went over all occurence of `TyAdt` and made sure `TyForeign` was always correctly handled. Inherent impls (which you had pointed out) and public-in-private checks were missing.

Hopefully there won't be any cases where `TyForeign` must be handled but not `TyAdt` I might have missed.

I didn't make a new adt kind because I've found that most of the time ADTs and extern types behave differently. There are only a few places where they are treated in the same way.

bors commented:
:umbrella: The latest upstream changes (presumably #44275) made this pull request unmergeable. Please resolve the merge conflicts.

arielb1 commented:
I hope this works well - now we just need a rebase.

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit 2120daf has been approved by `arielb1`

<!-- @bors r=arielb1 2120daf1d256841aa700a7f561abf9cb8ef57246 -->

arielb1 commented:
@bors r-

Why are you updating RLS?

arielb1 commented:
Nevermind it's ok - the RLS was updated by @nrc

@bors r+

bors commented:
:pushpin: Commit 2120daf has been approved by `arielb1`

<!-- @bors r=arielb1 2120daf1d256841aa700a7f561abf9cb8ef57246 -->

bors commented:
:hourglass: Testing commit 2120daf1d256841aa700a7f561abf9cb8ef57246 with merge db66bed1da778bafefee790e5cf26170a03d8832...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/274839679?utm_source=github_status&utm_medium=notification)

kennytm commented:
@bors retry #38618

bors commented:
:umbrella: The latest upstream changes (presumably #44420) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:lock: Merge conflict

plietar commented:
PR #44420 changed `rls` to its own branch as well. I've asked @nrc to rebase the extern-types branch on top of that.

bors commented:
:umbrella: The latest upstream changes (presumably #43017) made this pull request unmergeable. Please resolve the merge conflicts.

arielb1 commented:
status: waiting until RLS is updated?

aidanhs commented:
For people checking in for triage:
 - the commit RLS needs to be updated on top of is https://github.com/rust-lang-nursery/rls/commits/7221e38023c41ff2532ebbf54a7da296fd488b50
 - the RLS branch is https://github.com/rust-lang-nursery/rls/commits/rustfmt-extern-type

When the latter contains all of the commits of the former, this PR can move forward. Ping @nrc.

plietar commented:
I'm on vacation this week, I'll rebase and sort out rustfmt with @nrc when I get back.

plietar commented:
I've rebased and squashed.

I've marked rls and rustfmt as broken in `toolstate.toml`, which should allow this PR to land and rustfmt to be fixed asynchronously.

plietar commented:
Note that I've changed `<*const T>::is_null()` (and a few more methods in `core::ptr`) to work with `T: ?Sized`, since it is a very useful method in FFI contexts. This is instastable though, so it may need signoff from libs team.

eddyb commented:
@plietar IIRC, we don't have a sound way to create a NULL `*mut T` in general.
cc @rust-lang/lang - among other things, we *always assume* vtable pointers from fat pointers are not null, *even if* the pointer is raw.

plietar commented:
@eddyb I've only extended the `<*const T>::is_null()` and `<*const T>::as_ref()` methods, *not* `core::ptr::null()` specifically because of this. A pointer is null iff it's data part is null.

Currently you can create null pointers to extern types by casting 0 to a pointer, as I've added a special case to allow this, but not in a generic context.

It would be very nice to have a `fn null(): *const T` method that works for extern types as well, but there's no way to say `*const T` must be a thin pointer other than `T: Sized` which doesn't work here.

Alternatively, we could say that raw pointers make no guarantees about their metadata part, just like they make no guarantees about the data part, and allow null vtable pointers. Or say we allow null vtable pointers only if the data part is null.

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit da549e1 has been approved by `arielb1`

<!-- @bors r=arielb1 da549e116a335c3febbe40b72bb7deeca7039b2d -->

bors commented:
:hourglass: Testing commit da549e116a335c3febbe40b72bb7deeca7039b2d with merge 667de1dfdb152c0d81091a987dfe2cc94439f64f...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.4849)

plietar commented:
Based on the conversation in #44932 I've removed the changes to `core::ptr` which relaxes `is_null` to `!Sized` types. I'll leave it to that PR to figure out it's semantics and decide whether it is even desirable.

Unfortunately it seems like setting rustfmt/rls as "broken" doesn't actually work, so I'll have to fix these first. @nrc is currently on leave, and I've been too busy with other things to find and chase someone else, but I'll try and sort it out soon.

bors commented:
:umbrella: The latest upstream changes (presumably #44614) made this pull request unmergeable. Please resolve the merge conflicts.

arielb1 commented:
@plietar

So this is still waiting for rustfmt/rls? I think @jonathandturner can also help you get RLS fixed.

In any case, I'll be in the US next week for Rust Belt Rust and I probably won't be chasing notifications, so if I miss this, just ping me on IRC.

jonathandturner commented:
@arielb1 - I don't do much with dev tools these days.  You might want to ping other folks in @rust-lang/dev-tools

fitzgen commented:
I believe the RLS is mostly @nrc's domain now, however he is on PTO for a bit.

alexcrichton commented:
triage ping for @plietar!

I think you should be able to disable rustfmt/rls in `src/tools/toolstate.toml` now, and would that help along this PR?

oli-obk commented:
I don't think we have a strategy for producing nightlies without rls

arielb1 commented:
aturon had said that breaking rustfmt & rls should be ok. I'll rebase this when I have time.

plietar commented:
Hey,
Super sorry, I've been really busy and forgot about this. I've rebased now.
Now that rustbuild has support for broken rustfmt/rls this should be able to land.

kennytm commented:
r? @arielb1

arielb1 commented:
At last!
@bors r+

bors commented:
:pushpin: Commit 1e9e319 has been approved by `arielb1`

<!-- @bors r=arielb1 1e9e3191ab01ad16e8a839efeea5cac2999140af -->

bors commented:
:hourglass: Testing commit 1e9e3191ab01ad16e8a839efeea5cac2999140af with merge dce604a8fe85c22beb53d99c609d88e3a0ca9320...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5142), [status-travis](https://travis-ci.org/rust-lang/rust/builds/294117982?utm_source=github_status&utm_medium=notification)
Approved by: arielb1
Pushing dce604a8fe85c22beb53d99c609d88e3a0ca9320 to master...

eddyb commented:
Oh no, this shouldn't have merged before #45225. Oh well, time to rebase now.
